src/base/LruMap.h

```diff
@@ -135,11 +153,29 @@ LruMap<EntryValue, EntryCost>::get(const char *key)
     return NULL;
 }
 
-template <class EntryValue, size_t EntryCost>
+template <class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
 bool
-LruMap<EntryValue, EntryCost>::add(const char *key, EntryValue *t)
+LruMap<EntryValue, MemoryUsedByEV>::add(const char *key, EntryValue *t)
 {
-    if (ttl == 0)
+    // map the retention concept into a TTL value
+
+    // do not store / retain any entries
+    if (retention == 0)
+        return false;
+
+    // negative retention means store 'forever'
+    if (retention < 0)
+        return add(key, t, 0xFFFFFFFF);
+
+    // positive retention are TTL
+    return add(key, t, retention);
+}
+
+template <class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+bool
+LruMap<EntryValue, MemoryUsedByEV>::add(const char *key, EntryValue *t, int ttl)
+{
+    if (retention == 0)
```

Comment:
The why info varies for each template instantiator. This map simply supports it AFAIK. No reason was presented by the LruMap original author. I imagine it is because stale entries may still be actively used in some places (ie SSL certs for active connections) it does not imply useless nor invalid.

In the Bearer token use-case (which is out of scope for LruMap.h documentation) stale simply means a re-check is required before next use and discard is guaranteed to be safe. Fresh entries may be in-use so discard does not have the guarantee since it may lead to user annoyance with re-auth - safe in security terms, but a PITA for users.